### PR TITLE
On receiving an unexpected response, provide the reason in the error message of the returned status

### PR DESCRIPTION
**Pull request by y-zeng**
If streams are ended because the transport drops connection, optional_drop_message of the transport will be put in the error message. This message is set when:
- The received response is an http1.x response
- The received response is unparseable

Related issue: #4168

### CODE REVIEW

File: src/core/ext/transport/chttp2/transport/chttp2_transport.c [hunk 1]
```diff
@@ -1519,16 +1529,26 @@ static void close_from_api(grpc_exec_ctx *exec_ctx,
...
+  gpr_slice *optional_drop_message;
+} cancel_stream_cb_arg;
+
 static void cancel_stream_cb(grpc_chttp2_transport_global *transport_global,
                              void *user_data,
                              grpc_chttp2_stream_global *stream_global) {
-  cancel_from_api(user_data, transport_global, stream_global,
-                  GRPC_STATUS_UNAVAILABLE);
+  cancel_stream_cb_arg *arg = user_data;
+  cancel_from_api(arg->exec_ctx, transport_global, stream_global,
```
Comment (yang-g) [inline src/core/ext/transport/chttp2/transport/chttp2_transport.c]: Can you get the transport from transport_global here? Then you do not need to add the cancel_stream_cb_arg.

Comment (y-zeng) [inline src/core/ext/transport/chttp2/transport/chttp2_transport.c]: Fixed. Transport is obtained from transport_global using TRANSPORT_FROM_GLOBAL macro. Thanks for the review!

**Comment by yang-g** [comment]
LGTM
waiting for green.

**Comment by hsaliak** [comment]
@jtattermusch  could you please merge? /cc @yang-g

**Comment by ctiller** [comment]
This might need some further work... I doubt we're going to see a HTTP/1 response rolled up as a HTTP/2 message ever.

Recommend removing the HTTP/1 parsing part of this, and concentrating on pulling the HTTP status from the `:status` metadata (probably in http_client_filter.c).

File: src/core/ext/transport/chttp2/transport/chttp2_transport.c [hunk 2]
```diff
@@ -1610,6 +1647,14 @@ static void parsing_action(grpc_exec_ctx *exec_ctx, void *arg, bool success) {
     ;
   if (i != t->read_buffer.count) {
     success = false;
+    gpr_slice_unref(t->optional_drop_message);
+    if (try_http_parsing(exec_ctx, t)) {
```
Comment (ctiller) [inline src/core/ext/transport/chttp2/transport/chttp2_transport.c]: This is probably the wrong thing to do unless it's the first slice received on a connection?

**Pull request merged**